theano/gof/opt.py

```diff
@@ -533,19 +544,64 @@ def process_node(self, fgraph, node):
             if len(node.inputs) != len(candidate.inputs):
                 continue
 
+            cand_has_assert = False
+
+            # Get input list of the candidate with assert removed
+            cand_inputs_assert_removed = []
+            for i in candidate.inputs:
+                if i.owner and isinstance(i.owner.op,
+                                          theano.tensor.opt.Assert):
+                    cand_has_assert = True
+                    cand_inputs_assert_removed.append(i.owner.inputs[0])
+                else:
+                    cand_inputs_assert_removed.append(i)
+
+            # if the current node has assert input, it should not be
+            # replaced with a candidate node which has no assert input
```

Comment:
Yes it is OK to replace candidate. It will be new but OK.
Le 4 août 2015 01:57, "t13m" <notifications@github.com> a écrit :

> In theano/gof/opt.py
> <https://github.com/Theano/Theano/pull/3232#discussion_r36159459>:
>
> > @@ -533,19 +544,64 @@ def process_node(self, fgraph, node):
> >              if len(node.inputs) != len(candidate.inputs):
> >                  continue
> >
> > +            cand_has_assert = False
> > +
> > +            # Get input list of the candidate with assert removed
> > +            cand_inputs_assert_removed = []
> > +            for i in candidate.inputs:
> > +                if i.owner and isinstance(i.owner.op,
> > +                                          theano.tensor.opt.Assert):
> > +                    cand_has_assert = True
> > +                    cand_inputs_assert_removed.append(i.owner.inputs[0])
> > +                else:
> > +                    cand_inputs_assert_removed.append(i)
> > +
> > +            # if the current node has assert input, it should not be
> > +            # replaced with a candidate node which has no assert input
>
> I thought that in every iteration on node, only node was the one that
> should be replaced. So it is ok to replace the candidate?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3232/files#r36159459>.
>